src/mumble/widgets/TrayIcon.cpp

```diff
@@ -0,0 +1,192 @@
+		}
+	}
+
+	if (&newIcon.get() != &m_statusIcon.get()) {
+		m_statusIcon = newIcon;
+		applyIcon(m_statusIcon);
+	}
+}
+
+void TrayIcon::on_icon_clicked(QSystemTrayIcon::ActivationReason reason) {
+	switch (reason) {
+		case QSystemTrayIcon::Trigger:
+#ifndef Q_OS_MAC
+			// macOS is special as it both shows the context menu AND triggers the action.
+			// We only want at most one of those and since we can not prevent showing
+			// the menu, we skip the action.
+			toggleShowHide();
+#endif
+			break;
+		case QSystemTrayIcon::Unknown:
+		case QSystemTrayIcon::Context:
+		case QSystemTrayIcon::DoubleClick:
+		case QSystemTrayIcon::MiddleClick:
+			break;
+	}
+}
+
+void TrayIcon::updateContextMenu() {
+	m_contextMenu->clear();
+
+	if (Global::get().mw->isVisible() && !Global::get().mw->isMinimized()) {
+		m_hideAction->setEnabled(QSystemTrayIcon::isSystemTrayAvailable());
+		m_contextMenu->addAction(m_hideAction);
+	} else {
+		m_contextMenu->addAction(m_showAction);
+	}
+
+	m_contextMenu->addSeparator();
+
+	m_contextMenu->addAction(Global::get().mw->qaAudioMute);
+	m_contextMenu->addAction(Global::get().mw->qaAudioDeaf);
+	m_contextMenu->addAction(Global::get().mw->qaTalkingUIToggle);
+	m_contextMenu->addSeparator();
+	m_contextMenu->addAction(Global::get().mw->qaQuit);
+}
+
+void TrayIcon::applyIcon(QIcon &icon) {
+	setIcon(icon);
+	m_currentIcon = &icon;
+}
+
+void TrayIcon::toggleShowHide() {
+	if (Global::get().mw->isVisible() && !Global::get().mw->isMinimized()) {
+		on_hideAction_triggered();
+	} else {
+		on_showAction_triggered();
+	}
+}
+
+void TrayIcon::on_showAction_triggered() {
+	Global::get().mw->showRaiseWindow();
+	updateContextMenu();
+}
+
+void TrayIcon::on_hideAction_triggered() {
+	if (!QSystemTrayIcon::isSystemTrayAvailable()) {
+		// The system reports that no system tray is available.
+		// If we would hide Mumble now, there would be no way to
+		// get it back...
+		return;
+	}
+
+	if (qApp->activeModalWidget() || qApp->activePopupWidget()) {
+		// There is one or multiple modal or popup window(s) active, which
+		// would not be hidden by this call. So we also do not hide
+		// the MainWindow...
+		return;
+	}
+
+#ifndef Q_OS_MAC
+	Global::get().mw->hide();
+#else
+	// Qt can not hide the window via the native macOS hide function. This should be re-evaluated with new Qt versions.
+	// Instead we just minimize.
+	Global::get().mw->setWindowState(Global::get().mw->windowState() | Qt::WindowMinimized);
+#endif
+
+	updateContextMenu();
+}
+
+void TrayIcon::highlight() {
+	on_timer_triggered();
+}
+
+void TrayIcon::unhighlight() {
+	if (m_highlightTimer == nullptr || !m_highlightTimer->isActive()) {
+		return;
+	}
+
+	m_highlightTimer->stop();
+	applyIcon(m_statusIcon);
+}
+
+void TrayIcon::on_timer_triggered() {
+	if (m_currentIcon == &m_statusIcon.get()) {
+		applyIcon(Global::get().mw->m_iconInformation);
+		m_highlightTimer->start(500);
+	} else {
+		applyIcon(m_statusIcon);
+		m_highlightTimer->start(2000);
+	}
```

Comment:
What is this logic for? I think this should have a comment explaining what is to be achieved here.